src/tpm2/crypto/openssl/CryptSym.c

```diff
@@ -458,6 +461,262 @@ CryptSymmetricDecrypt(
+    int                     i;
+    BYTE                    tmp[MAX_SYM_BLOCK_SIZE];
+    BYTE                   *pT;
+
+    TDES_set_encrypt_key(key, keySizeInBits,
+                         (tpmKeyScheduleTDES *)&keySchedule.TDES);
+
+    for(; dSize > 0; dSize -= blockSize)
+	{
+	    // Encrypt the current value of the IV(counter)
+	    TDES_encrypt(iv, tmp, (tpmKeyScheduleTDES *)&keySchedule.TDES);
+	    //increment the counter (counter is big-endian so start at end)
+	    for(i = blockSize - 1; i >= 0; i--)
+		if((iv[i] += 1) != 0)
+		    break;
+	    // XOR the encrypted counter value with input and put into output
+	    pT = tmp;
+	    for(i = (dSize < blockSize) ? dSize : blockSize; i > 0; i--)
+		*dOut++ = *dIn++ ^ *pT++;
+	}
+}
+#endif
+
+/* 10.2.20.5 Symmetric Encryption */
+/* This function performs symmetric encryption based on the mode. */
+/* Error Returns Meaning */
+/* TPM_RC_SIZE dSize is not a multiple of the block size for an algorithm that requires it */
+/* TPM_RC_FAILURE Fatal error */
+LIB_EXPORT TPM_RC
+CryptSymmetricEncrypt(
+		      BYTE                *dOut,          // OUT:
+		      TPM_ALG_ID           algorithm,     // IN: the symmetric algorithm
+		      UINT16               keySizeInBits, // IN: key size in bits
+		      const BYTE          *key,           // IN: key buffer. The size of this buffer
+		      //     in bytes is (keySizeInBits + 7) / 8
+		      TPM2B_IV            *ivInOut,       // IN/OUT: IV for decryption.
+		      TPM_ALG_ID           mode,          // IN: Mode to use
+		      INT32                dSize,         // IN: data size (may need to be a
+		      //     multiple of the blockSize)
+		      const BYTE          *dIn            // IN: data buffer
+		      )
+{
+    INT16                blockSize;
+    BYTE                *iv;
+    BYTE                 defaultIv[MAX_SYM_BLOCK_SIZE] = {0};
+    evpfunc              evpfn;
+    EVP_CIPHER_CTX      *ctx = NULL;
+    int                  outlen1, outlen2;
+    BYTE                *pOut = dOut;
+    BYTE                *buffer = NULL; // for in-place encryption
+    UINT32               buffersize = 0;
+    BYTE                 keyToUse[MAX_SYM_KEY_BYTES];
+    UINT16               keyToUseLen = (UINT16)sizeof(keyToUse);
+    TPM_RC               retVal = TPM_RC_SUCCESS;
+
+    pAssert(dOut != NULL && key != NULL && dIn != NULL);
+    if(dSize == 0)
+	return TPM_RC_SUCCESS;
+    TEST(algorithm);
+    blockSize = CryptGetSymmetricBlockSize(algorithm, keySizeInBits);
+    if(blockSize == 0)
+	return TPM_RC_FAILURE;
+    // If the iv is provided, then it is expected to be block sized. In some cases,
+    // the caller is providing an array of 0's that is equal to [MAX_SYM_BLOCK_SIZE]
+    // with no knowledge of the actual block size. This function will set it.
+    if((ivInOut != NULL) && (mode != ALG_ECB_VALUE))
+	{
+	    ivInOut->t.size = blockSize;
+	    iv = ivInOut->t.buffer;
+	}
+    else
+	iv = defaultIv;
```

Comment:
If there is no IV in input, shouldn't this interface generate an IV and return it ?
ivInOut is marked as IN/OUT but I do not see an IV is ever returned

Reply:
In most cases the IV is a variable on the stack and passed into this function and then not looked at anymore. In some cases the IV comes in via a TPM command directly and is in the input parameters, in which cases it's also not looked at anymore. So I think this IN/OUT is an OUT mostly for efficiency purposes because the buffer of the IV may get changed but seems to not be of interest to the caller afterwards.

Reply:
... and regarding generation of the IV: I am using the same code for the defaultIV as the original code does further up.